doc/developer/design/20241008_expression_cache.md

```diff
@@ -0,0 +1,243 @@
+    deploy_generation: u64,
+    information_needed_to_connect_to_durable_store: _,
+}
+
+impl ExpressionCache {
+    /// Creates a new [`ExpressionCache`] for `deploy_generation`.
+    fn new(&mut self, deploy_generation: u64, information_needed_to_connect_to_durable_store: _) -> Self;
+
+    /// Reconciles all entries in current deploy generation with the current objects, `current_ids`,
+    /// and current optimizer features, `optimizer_features`.
+    ///
+    /// If `remove_prior_gens` is `true`, all previous generations are durably removed from the
+    /// cache.
+    ///
+    /// Returns all cached expressions in the current deploy generation, after reconciliation.
+    fn open(&mut self, current_ids: &BTreeSet<GlobalId>, optimizer_features: &OptimizerFeatures, remove_prior_gens: bool) -> Vec<(GlobalId, Expressions)>;
+
+    /// Durably inserts `expressions` into current deploy generation. This may also invalidate
+    /// entries giving by `expressions`.
+    ///
+    /// Returns a [`Future`] that completes once the changes have been made durable.
+    ///
+    /// Panics if any `GlobalId` already exists in the cache.
+    fn insert_expressions(&mut self, expressions: Vec<NewEntry>) -> impl Future<Output=()>;
+
+    /// Durably remove and return all entries in current deploy generation that depend on an ID in
+    /// `dropped_ids` .
+    ///
+    /// Optional for v1.
+    fn invalidate_entries(&mut self, dropped_ids: BTreeSet<GlobalId>) -> Vec<(GlobalId, Expressions)>;
+}
+```
+
+### Startup
+
+Below is a detailed set of steps that will happen in startup.
+
+1. Call `ExpressionCache::open` to read the cache into memory and perform reconciliation (See
+   [Startup Reconciliation](#startup reconciliation)). When passing in the arguments,
+   `remove_prior_gens == !read_only_mode`.
+2. While opening the catalog, for each object:
+    a. If the object is present in the cache, use cached optimized expression.
+    b. Else generate the optimized expressions and insert the expressions via
+       `ExpressionCache::insert_expressions`. This will also perform any necessary invalidations if
+       the new expression is an index. See ([Create Invalidations](#create invalidations)).
+
+#### Startup Reconciliation
+When opening the cache for the first time, we need to perform the following reconciliation tasks:
+
+  - Remove any entries that exist in the cache but not in the catalog.
+  - If `remove_prior_gens` is true, then remove all prior gens.
+
+
+### DDL - Create
+
+1. Execute catalog transaction.
+2. Update cache via `ExpressionCache::insert_expressions`.
+
+#### Create Invalidations
+
+When creating and inserting a new index, we need to invalidate some entries that may optimize to
+new expressions. When creating index `i` on object `o`, we need to invalidate the following objects:
+
+  - `o`.
+  - All compute objects that depend directly on `o`.
+  - All compute objects that would directly depend on `o`, if all views were inlined.
```

Comment:
There should be an optional "re-compute and repopulate cache entries that could use the new index" step here, yeah?

Reply:
Good point, added.